pkg/results/reader.go

```diff
@@ -0,0 +1,227 @@
+	r := &Reader{
+		Reader: reader,
+	}
+
+	conf := &config.Config{}
+
+	err := r.WalkFiles(func(path string, info os.FileInfo, err error) error {
+		return ExtractConfig(path, info, conf)
+	})
+	if err != nil {
+		return "", err
+	}
+	var version string
+	// Get rid of any of the extra version information that doesn't affect archive layout.
+	// Example: v0.10.0-a2b3d4
+	if strings.HasPrefix(conf.Version, versionEight) {
+		version = versionEight
+	} else if strings.HasPrefix(conf.Version, versionNine) {
+		version = versionNine
+	} else if strings.HasPrefix(conf.Version, versionTen) {
+		version = versionTen
+	} else {
+		return "", errors.New("cannot discover Sonobuoy archive version")
+	}
+	return version, nil
+}
+
+// tarFileInfo implements os.FileInfo and extends the Sys() method to
+// return a reader to a file in a tar archive.
+type tarFileInfo struct {
+	os.FileInfo
+	io.Reader
+}
+
+// Sys is going to be an io.Reader to a file in a tar archive.
+// This is how data is extracted from the archive.
+func (t *tarFileInfo) Sys() interface{} {
+	return t.Reader
+}
+
+// WalkFiles walks all of the files in the archive.
+func (r *Reader) WalkFiles(walkfn filepath.WalkFunc) error {
+	tr := tar.NewReader(r)
+	var err error
+	var header *tar.Header
+	for {
+		header, err = tr.Next()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return err
+		}
+		info := &tarFileInfo{
+			header.FileInfo(),
+			tr,
+		}
+		err = walkfn(filepath.Clean(header.Name), info, err)
+	}
+	return nil
+
+}
+
+// Functions to be used within a walkfn
+
+// ExtractIntoStruct takes a predicate function and some file information
+// and decodes the contents of the file that matches the predicate into the
+// interface passed in (generally a pointer to a struct/slice).
+func ExtractIntoStruct(predicate func(string) bool, path string, info os.FileInfo, object interface{}) error {
+	if predicate(path) {
+		reader, ok := info.Sys().(io.Reader)
+		if !ok {
+			return errors.New("info.Sys() is not a reader")
+		}
+		decoder := json.NewDecoder(reader)
+		err := decoder.Decode(object)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// ExtractFileIntoStruct is a helper for a common use case of extracting
+// the contents of one file into the object.
+func ExtractFileIntoStruct(file, path string, info os.FileInfo, object interface{}) error {
+	return ExtractIntoStruct(func(p string) bool {
+		return file == p
+	}, path, info, object)
+}
+
+// ExtractConfig populates the config object regardless of version.
+func ExtractConfig(path string, info os.FileInfo, conf *config.Config) error {
+	return ExtractIntoStruct(func(file string) bool {
+		return path == ConfigFile(versionTen) || path == ConfigFile(versionEight)
+	}, path, info, conf)
+}
+
+// Functions for helping with backwards compatibility
+
+func (a *Reader) ServerVersionFile() string {
+	switch a.Version {
+	case versionEight:
+		return "serverversion/serverversion.json"
```

Comment:
should maybe be a constant to?

Reply:
yeah, I'll revisit these in future versions